mailpoet/lib/Subscribers/SubscribersEmailCountsController.php

```diff
@@ -0,0 +1,142 @@
+          JOIN {$scheduledTasksTable} as st ON sq.task_id = st.id
+          WHERE st.processed_at IS NOT NULL
+          AND st.processed_at < :dayAgo";
+      $queryParams['dayAgo'] = $dayAgoIso;
+
+      if ($dateLastProcessed) {
+        $processedTaskIdsTableSql .= " AND st.processed_at >= :dateFrom";
+        $carbonDateLastProcessed = Carbon::createFromTimestamp($dateLastProcessed->getTimestamp());
+        $dateFromIso = ($carbonDateLastProcessed->subDay())->toDateTimeString();
+        $queryParams['dateFrom'] = $dateFromIso;
+      }
+
+      $resultQuery = $connection->executeQuery($processedTaskIdsTableSql, $queryParams);
+      $this->processedTaskIdsTableCreated = true;
+
+      if ($resultQuery->rowCount() === 0) return [0,0];
+    }
+
+    // Temporary table needed for UPDATE query
+    // mySQL does not allow to modify the same table used in the select
+    $subscriberIdsEmailsCountTmpTable = 'subscribers_ids_email_counts';
+    $connection->executeQuery("
+      CREATE TEMPORARY TABLE IF NOT EXISTS {$subscriberIdsEmailsCountTmpTable}
+      (UNIQUE subscriber_id (id))
+        SELECT s.id, count(task_ids.id) as emails_count from {$this->subscribersTable} s
+        JOIN {$scheduledTaskSubscribersTable} sts ON s.id = sts.subscriber_id
+        JOIN {$processedTaskIdsTable} task_ids ON task_ids.id = sts.task_id
+        WHERE s.id >= :startId
+        AND s.id <= :endId
+        GROUP BY s.id
+    ",
+      [
+        'startId' => $startId,
+        'endId' => $endId,
+      ]
+    );
+
+    // If $dateLastProcessed provided, increment value, otherwise count all and reset value
+    $initUpdateValue = $dateLastProcessed ? 's.emails_count' : '';
+    $updateQuery = $connection->executeQuery("
+      UPDATE {$this->subscribersTable} as s
+      JOIN {$subscriberIdsEmailsCountTmpTable} as sc ON s.id = sc.id
+      SET s.emails_count = {$initUpdateValue} + IFNULL(sc.emails_count, 0)
+      WHERE s.id >= :startId
+      AND s.id <= :endId
+    ",
+      [
+        'startId' => $startId,
+        'endId' => $endId,
+      ]
+    );
+    $connection->executeQuery("DROP TABLE {$subscriberIdsEmailsCountTmpTable}");
+
+    return [$countSubscribersToUpdate, $endId];
+  }
+
+  private function countAndMaxOfSubscribersInRange(int $startId, int $batchSize): array {
+    $connection = $this->entityManager->getConnection();
+
+    $result = $connection->executeQuery("
+      SELECT s.id FROM {$this->subscribersTable} as s
+      WHERE s.id >= :startId
+      ORDER BY s.id
+      LIMIT :batchSize
+    ",
+      [
+        'startId' => $startId,
+        'batchSize' => $batchSize,
+      ],
+      [
+        'startId' => \PDO::PARAM_INT,
+        'batchSize' => \PDO::PARAM_INT,
+      ]
+    );
+
+    $subscribersInRange = $result->fetchAllAssociative();
+
+    $countSubscribersInRange = count(array_map(
+      function ($id) {
+        return (int)$id['id'];
+      },
+      $subscribersInRange
+    ));
```

Comment:
`SELECT MAX(s.id) AS max, COUNT(s.id) as count` would not work. 
I am using `LIMIT` in the query so that, if there are gaps in the subscribers IDs, we don't run an empty loop.
MAX(s.id) returns the max of the whole table and I need the MAX id of current batch.

Reply:
Oh, right, that's true. But still I think it's true that `count($subscribersInRange)` would give the same result as the whole `count(array_map(...))` so I think the `array_map` is unnecessary here.

Or you could use a subquery:
```sql
SELECT COALESCE(MAX(ids.id), 0) AS max, COUNT(ids.id) as count FROM (
  SELECT s.id FROM wp_mailpoet_subscribers as s
  WHERE s.id >= 0
  ORDER BY s.id
  LIMIT 10
) ids
```

Reply:
Thanks for the suggestions. Modified to use a subquery.